feat: pass list of browsers to plugins file

bahmutov opened the pull request:
Passes list of detected browsers to the plugins file and lets the user modify the list.

- closes #5067 
- closes #5712

### User facing changelog

- Cypress now includes the full list of browsers found and their properties within the Cypress configuration. This means that the browsers array is also editable  within plugins by accessing `config.browsers`. The currently available browsers is also displayed within the configuration panel under Settings in the Test Runner.

### Additional details

Reworking browser detection and config to do the following:

1. find all browsers
2. set the browsers on the config object
3. pass the config to the plugins if any, allowing the user to adjust the list and return the updated list
4. if running in the interactive mode (`cypress open`)
  - show the updated list in the browser picker
5. else pick the browser from the updated list.
  - by default, it is `electron`, which if the user removes from the list will cause an error
  - if the user passed `--browser <name>` CLI then pick that browser from the current updated list of browsers
  - if the user passed `--browser <path>` disregard the updated list and discover the browser just like we do now

### How has the user experience changed?

Example use: https://github.com/cypress-io/cypress-example-electron/blob/master/cypress/plugins/index.js

#### Config Before

<img width="801" alt="Screen Shot 2019-11-14 at 9 26 24 AM" src="https://user-images.githubusercontent.com/1271364/68865474-e002f880-06c0-11ea-96b0-56bd0e3ec970.png">


#### Config After

![2019-11-13_16-43-09](https://user-images.githubusercontent.com/1578436/68806937-e3e93900-0634-11ea-8128-4351cfefa009.gif)

### PR Tasks

<!-- The following tasks must be completed before a PR can be merged.
You can delete tasks if they are not applicable to the PR changes. -->

- [x] Have tests been added/updated for the changes in this PR?
- [x] Has a PR to [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation) been submitted to document any user-facing changes? https://github.com/cypress-io/cypress-documentation/pull/2245
- [x] Has the original issue been tagged with a release in ZenHub? <!-- (internal team only)-->
- [x] update resolved configuration display, right now shows `[object], ...` for the list of browsers
- [ ] add a test to make sure user cannot list browsers in `cypress.json` but in the plugins only

bahmutov committed:
WIP: pass list of browsers to plugins file

bahmutov commented:
Testing it locally with command
```
DEBUG=cypress:launcher,cypress:server:plugins:child npm run dev -- --project ~/git/cypress-example-electron/
```

Which shows the browser detection happens first, and then the plugins file loading, so we do have this information

```
  cypress:launcher checking one browser canary +0ms
  cypress:launcher looking up canary on darwin platform +0ms
  cypress:launcher looking for app Contents/MacOS/Google Chrome Canary id com.google.Chrome.canary +0ms
  cypress:launcher looking for bundle id com.google.Chrome.canary using command: mdfind 'kMDItemCFBundleIdentifier=="com.google.Chrome.canary"' | head -1 +0ms
  cypress:launcher found com.google.Chrome.canary at /Applications/Google Chrome Canary.app +49ms
  cypress:launcher reading property file "/Applications/Google Chrome Canary.app/Contents/Info.plist" +1ms
  cypress:launcher setting major version for {"name":"canary","family":"chrome","displayName":"Canary","version":"78.0.3896.6","path":"/Applications/Google Chrome Canary.app/Contents/MacOS/Google Chrome Canary"} +17ms
  cypress:launcher browser canary version 78.0.3896.6 major version 78 +0ms
  cypress:server:plugins:child pluginsFile: /Users/gleb/git/cypress-example-electron/cypress/plugins/index.js +0ms
  cypress:server:plugins:child require pluginsFile +2ms
/Users/gleb/git/cypress-example-electron/cypress/plugins/index.js
[ '/Users/gleb/git/cypress/packages/electron/dist/Cypress/Cypress.app/Contents/Frameworks/Cypress Helper.app/Contents/MacOS/Cypress Helper',
  '/Users/gleb/git/cypress/packages/server/lib/plugins/child/index.js',
  '--file',
  '/Users/gleb/git/cypress-example-electron/cypress/plugins/index.js' ]
  cypress:server:plugins:child plugins load file "/Users/gleb/git/cypress-example-electron/cypress/plugins/index.js" +14ms
  cypress:server:plugins:child run plugins function +1ms
  cypress:server:plugins:child register event _get:task:body with id 0 +0ms
  cypress:server:plugins:child register event _get:task:keys with id 1 +0ms
argument config
{ animationDistanceThreshold: 5,
```

bahmutov committed:
pass browsers to the plugins and use returned list

bahmutov committed:
fix server unit tests

flotwig commented on packages/server/lib/config.coffee:
```diff
@@ -327,9 +340,24 @@ module.exports = {
+    # do not allow user to delete "browsers" list - otherwise how to run tests?
+    if Array.isArray(userBrowserList)
+      debug("using user supplied list of browsers %o", userBrowserList)
+      resolved.browsers = userBrowserList
```
Doesn't line 346 copy `diffs.browsers` in to `resolved` unconditionally?

bahmutov replied:
right, the default behavior is to use all browsers. But if the user has returned a list of browsers, we need to resolve with that list

flotwig commented on packages/server/lib/config.coffee:
```diff
@@ -327,9 +340,24 @@ module.exports = {
+    ## modifes detected browsers, the _.defaultsDeep merges it back
+    ## and we need to use the user's value
+
+    # do not allow user to delete "browsers" list - otherwise how to run tests?
```
a test for this would be nice

bahmutov replied:
tests would be nice in general

bahmutov commented on packages/server/lib/config.coffee:
```diff
@@ -309,6 +319,9 @@ module.exports = {
     diffs = deepDiff(cfg, overrides, true)
+    debug("config diffs %o", diffs)
+
+    userBrowserList = diffs && diffs.browsers && R.clone(diffs.browsers)
```
Yes, `_.clone` is shallow, you have to use `_.cloneDeep` to clone the entire object. `R.clone` being sane clones deep by default

bahmutov committed:
Merge branch 'develop' into pass-list-of-browsers-to-config-5067

bahmutov committed:
Merge branch 'develop' into pass-list-of-browsers-to-config-5067

jennifer-shehane reviewed (dismissed):
In terms of the API changes, @brian-mann mentioned that we should have a method call like `setBrowser` or `setBrowsers` so that we could actually do some validation on what is passed in - instead of the static overriding of the object. A user could mutate it and just completely mess up the launching of Cypress browser and we want to try and avoid that situation if possible.

jennifer-shehane committed:
Merge branch 'develop' into pass-list-of-browsers-to-config-5067

bahmutov committed:
Merge branch 'develop' into pass-list-of-browsers-to-config-5067

bahmutov commented:
I think simplicity right now wins - if you mutate the other configuration values, you can screw things up, but how often users would screw the list of browsers and leave the invalid configuration? If they send an invalid browser, we could validate - just like we should validate the entire modified config object after getting it back.

bahmutov commented:
This PR also enables to run tests in Microsft Edge Beta (Chromium-based) or Brave browser by simply returning them from the plugins file

```js
module.exports = (on, config) => {
  // `on` is used to hook into various events Cypress emits
  // `config` is the resolved Cypress config

  const pathToBrave =
    '/Applications/Brave Browser.app/Contents/MacOS/Brave Browser'
  config.browsers.push({
    name: 'brave',
    family: 'chrome',
    displayName: 'Brave',
    version: 'unknown',
    path: pathToBrave,
    majorVersion: 'unknown'
  })

  return config
}
```

<img width="1199" alt="Screen Shot 2019-10-19 at 8 09 23 AM" src="https://user-images.githubusercontent.com/2212006/67144756-6e15cb80-f248-11e9-813d-1e7ec55d1e71.png">

bahmutov committed:
Merge branch 'develop' into pass-list-of-browsers-to-config-5067

bahmutov committed:
Merge branch 'develop' into pass-list-of-browsers-to-config-5067

bahmutov committed:
adding returned list of browsers validation

bahmutov committed:
Merge branch 'develop' into pass-list-of-browsers-to-config-5067

bahmutov committed:
Merge branch 'develop' into pass-list-of-browsers-to-config-5067

bahmutov committed:
Merge branch 'develop' into pass-list-of-browsers-to-config-5067

bahmutov committed:
Merge branch 'pass-list-of-browsers-to-config-5067' of github.com:cypress-io/cypress into pass-list-of-browsers-to-config-5067

bahmutov committed:
Merge branch 'develop' into pass-list-of-browsers-to-config-5067

bahmutov committed:
Merge branch 'develop' into pass-list-of-browsers-to-config-5067

bahmutov committed:
Merge branch 'pass-list-of-browsers-to-config-5067' of github.com:cypress-io/cypress into pass-list-of-browsers-to-config-5067

bahmutov committed:
worked more on browser list validation

bahmutov committed:
update logic to only check if passed browsers

bahmutov committed:
wip: updating unit tests

bahmutov committed:
add more tests

bahmutov committed:
update integration tests

bahmutov committed:
add browsers list to the configuration spec

bahmutov committed:
update blacklist hosts display test

bahmutov committed:
Merge branch 'develop' into pass-list-of-browsers-to-config-5067

bahmutov commented:
Updated list of browsers in the config view

<img width="1074" alt="Screen Shot 2019-11-06 at 1 10 00 PM" src="https://user-images.githubusercontent.com/2212006/68325115-cfd29400-0096-11ea-894d-47e5d242b32b.png">

bahmutov committed:
resolve browsers by default

bahmutov committed:
Merge branch 'pass-list-of-browsers-to-config-5067' of github.com:cypress-io/cypress into pass-list-of-browsers-to-config-5067

bahmutov committed:
do not set browsers as CLI by default

bahmutov committed:
update code to not fail all tests when the list of browsers is the default empty one

bahmutov committed:
Merge branch 'develop' into pass-list-of-browsers-to-config-5067

bahmutov committed:
debug log:

bahmutov committed:
Merge branch 'pass-list-of-browsers-to-config-5067' of github.com:cypress-io/cypress into pass-list-of-browsers-to-config-5067

bahmutov committed:
detect browsers first when running project

bahmutov committed:
if passed browser, only search in the browsers resolved by the plugins

bahmutov committed:
allow browser path to be any string

bahmutov committed:
Merge branch 'develop' into pass-list-of-browsers-to-config-5067

bahmutov committed:
update integration test

bahmutov committed:
Merge branch 'pass-list-of-browsers-to-config-5067' of github.com:cypress-io/cypress into pass-list-of-browsers-to-config-5067

bahmutov committed:
Merge branch 'develop' into pass-list-of-browsers-to-config-5067

bahmutov committed:
Merge branch 'develop' into pass-list-of-browsers-to-config-5067

bahmutov committed:
add e2e test for filtering browsers in the plugins file

bahmutov committed:
Merge branch 'develop' into pass-list-of-browsers-to-config-5067

bahmutov committed:
allow browser major to be a number

bahmutov committed:
parse majorVersion for detected browser

flotwig commented on packages/server/__snapshots__/3_plugins_spec.coffee.js:
```diff
@@ -374,3 +374,12 @@ exports['e2e plugins calls after:screenshot for cy.screenshot() and failure scre
+
+Browser: 'chrome' was not found on your system.
+
+Available browsers found are: electron
```
neat, love that this can be programmatically controlled now

flotwig commented on packages/launcher/lib/detect.ts:
```diff
@@ -15,7 +15,14 @@ import {
+}
+
+// TODO: make this function NOT change its argument
+export const setMajorVersion = <T extends HasVersion>(browser: T): T => {
```
big ups for cleaning up the types here

flotwig commented on packages/launcher/lib/detect.ts:
```diff
@@ -15,7 +15,14 @@ import {
+  version?: string
+  majorVersion?: string | number
+  name: string
+}
```
```ts
type HasVersion = Pick<FoundBrowser, 'version' | 'majorVersion' | 'name'>
```

Does our typescript version support `Pick`? Would be easier to maintain

bahmutov replied:
I tried using Pick - but it messed up the return type :(

flotwig commented on packages/server/lib/browsers/index.coffee:
```diff
@@ -93,7 +100,9 @@ module.exports = {
   close: kill
 
   getAllBrowsersWith: (nameOrPath) ->
+    debug("getAllBrowsersWith")
```
```suggestion
    debug("getAllBrowsersWith %o", { nameOrPath })
```

flotwig commented on packages/server/lib/browsers/index.coffee:
```diff
@@ -93,7 +100,9 @@ module.exports = {
   getAllBrowsersWith: (nameOrPath) ->
+    debug("getAllBrowsersWith")
     if nameOrPath
+      debug("by name or path %s", nameOrPath)
```
```suggestion
```

flotwig commented on packages/server/lib/browsers/utils.coffee:
```diff
@@ -1,9 +1,11 @@
 path     = require("path")
+debug    = require("debug")("cypress:server:browsers")
```
somehow, some of these files have a 4th tag section and some don't... for the sake of eventual consistency, can this be used instead?

```suggestion
debug    = require("debug")("cypress:server:browsers:utils")
```

flotwig commented on packages/server/lib/browsers/utils.coffee:
```diff
@@ -85,18 +87,23 @@ module.exports = {
   getBrowsers: ->
     ## TODO: accept an options object which
     ## turns off getting electron browser?
```
The Electron browser can be removed, right? So this `TODO` can be struck from the code?

flotwig commented on packages/server/lib/config.coffee:
```diff
@@ -204,13 +216,45 @@ validateFile = (file) ->
+  # use try / catch inside to make sure we report the error
+  # and don't swallow it during integration testing
+  # otherwise I had situations where a test was failing in a very weird
+  # ways because an error was thrown here but the test continued
```
curious about this, which spec was failing?

bahmutov replied:
don't remember - just remember silence, might be because we were catching errors in tests, but I want this error to be discoverable

flotwig replied:
i don't see how `throw err` differs from `try { throw err } catch (err) { throw err }`, are you 100% sure this is necessary?

bahmutov replied:
it is NOT throwing the error - the error propagates. It is that we don't see that the invalid browser was the source of error, which allows one to chase wild guesses, rather than the true reason for this error

flotwig replied:
ah, so this attaches the proper stack trace?

bahmutov replied:
no, it is still the same, I believe - but this prints the error before throwing it again, allowing you to see it

flotwig replied:
must've missed that debug statement. got it, thanks!

flotwig commented on packages/server/lib/open_project.coffee:
```diff
@@ -209,10 +212,11 @@ create = ->
       ## open the project and return
       ## the config for the project instance
       debug("opening project %s", path)
+      debug("and options %o", options)
```
```js
debug('opening project %o', { path, options }')
```

bahmutov replied:
so I split it across two lines because options is a large object, which kind of hides the path. By having these two on separate lines I get nice stdout that's easy to scan

flotwig reviewed (changes requested):
very nice

brian-mann reviewed (changes requested):
Make arrays of objects collapsed by default in the `Settings` UI. @andrew-codes maybe ;-)

bahmutov committed:
Update packages/server/lib/browsers/utils.coffee

Co-Authored-By: Zach Bloomquist <github@chary.us>

bahmutov committed:
Update packages/server/lib/browsers/index.coffee

Co-Authored-By: Zach Bloomquist <github@chary.us>

bahmutov committed:
Update packages/server/lib/browsers/index.coffee

Co-Authored-By: Zach Bloomquist <github@chary.us>

bahmutov committed:
remove unnecessary TODO

bahmutov committed:
convert major version for Electron browser

bahmutov commented:
When we show resolved project settings using packages/desktop-gui/src/settings/configuration.jsx we show expanded list of browsers which is a lot of objects, typically. It would be nice to collapse this list by default and allow user to expand it by clicking

<img width="1074" alt="Screen Shot 2019-11-12 at 4 04 01 PM" src="https://user-images.githubusercontent.com/2212006/68710747-20019880-0566-11ea-97d7-371be23d248d.png">

I think in the current configuration, `browsers` is the only array of objects, so if we collapse arrays of objects by default, we should be fine @andrew-codes

bahmutov committed:
update Electron port in integration test

bahmutov committed:
change config spec to expect a number

flotwig reviewed (dismissed):
lgtm once the browser-list expando is added. there's plenty of libraries that we could drop in to do this, like https://github.com/storybookjs/react-inspector

andrew-codes commented:
@flotwig if you have other suggestions also, I'm open to them. I looked at [storybookjs/react-inspector](https://github.com/storybookjs/react-inspector), but I am unsure if it will enable us to color code or provide access to enable tooltips for the origin of the setting's value.

flotwig commented:
> I am unsure if it will enable us to color code or provide access to enable tooltips for the origin of the setting's value.

@andrew-codes I think you can pass a custom `nodeRenderer`, which will allow you to customize the rendering of each node of the object. Maybe that will give the flexibility we need.

andrew-codes committed:
browsers values are collapsed by default

Provide a button to "see more..." (see browser data). Once clicked, browser data is shown and button changes to "see less...", which will collpase browser data.

andrew-codes committed:
use react-inspector for config viewer

andrew-codes commented:
@brian-mann How does this look to you?

![2019-11-13_16-43-09](https://user-images.githubusercontent.com/1578436/68806937-e3e93900-0634-11ea-8128-4351cfefa009.gif)

andrew-codes committed:
Merge branch 'develop' into pass-list-of-browsers-to-config-5067

brian-mann commented:
Can we get rid of the root `config` object? It's super weird.

Need to work on the colors - I get that it's trying to mimic the dev tools, but the purple key values are way too strong and overwhelm the others - and the whitespace for the strings need work. 

I also think that the strings need actual quotes around them like before to help indicate what kind of type it is.

When you expand the nested objects, can you hide the preview? It's weird that it stays sticking out to the right.

brian-mann reviewed (changes requested):
Can we get rid of the root config object? It's super weird.

Need to work on the colors - I get that it's trying to mimic the dev tools, but the purple key values are way too strong and overwhelm the others - and the whitespace for the strings need work.

I also think that the strings need actual quotes around them like before to help indicate what kind of type it is.

When you expand the nested objects, can you hide the preview? It's weird that it stays sticking out to the right.

jennifer-shehane commented:
@brian-mann I think the colors are fine. It still maintains good usability. If anything - this may be a lot easier to read than before and it's still very  quick to find the config and where. 

Agree about the base `"{}"` config string being there - I think this may be some old artifact carried over.

jennifer-shehane commented:
@bahmutov Can you open a PR in the documentation for these changes since the API is solid now? If there is one and I missed it - then tag it in the main PR comment. Thanks.

bahmutov commented:
Hmm, for some reason the configuration shows the list of browsers coming from `default` settings, but this should be coming from the `plugins`.

<img width="908" alt="Screen Shot 2019-11-14 at 10 17 47 AM" src="https://user-images.githubusercontent.com/2212006/68870051-30ca1f80-06c8-11ea-9924-5507ef6b07fc.png">

andrew-codes committed:
address PR feedback

- hide summary of keys once value is expanded
- string values are now quoted string values
- from coloring now correctly applies and does so to all expanded children

andrew-codes commented:
@bahmutov this has now been addressed, along with some of @brian-mann 's feedback.

![2019-11-14_14-02-19](https://user-images.githubusercontent.com/1578436/68887753-89f57b80-06e7-11ea-960a-50d1797400ec.gif)

andrew-codes committed:
account for slight delay of rendering the config

andrew-codes committed:
correct test assertion

Chrome:0 should be found in the case that browsers have been expanded

brian-mann reviewed (changes requested):
Much better... I see 3 little issues:

1. Is there any way to remove the root `config` key that everything is nested under? I'm worried this is going to be confusing... could we just add back in the curly braces and wrap those around the top and bottom? Or just remove the root `config` altogether if that's easier.
2. I notice that `undefined` is showing up as a string: `"env": "undefined"`
3. Can we add margin-left to the values? When the values are highlighted in a background color they touch the colon `:` and makes it feel super squished.

andrew-codes committed:
additional PR feedback

- more spacing between items
- do not show the root config
- do not quote null/undefined values

andrew-codes commented:
@brian-mann With those suggestions:

![image](https://user-images.githubusercontent.com/1578436/68895134-bc5aa500-06f6-11ea-8f92-c363c20748e1.png)

andrew-codes committed:
missing color on plugins legend item

andrew-codes committed:
Merge branch 'develop' into pass-list-of-browsers-to-config-5067

andrew-codes committed:
Wrap config in curly braces & style updates

brian-mann commented on packages/launcher/lib/detect.ts:
```diff
@@ -24,6 +31,10 @@ const setMajorVersion = (browser: FoundBrowser) => {
     )
+
+    if (browser.majorVersion) {
+      browser.majorVersion = parseInt(browser.majorVersion)
```
you pretty much always want `parseFloat` here or `Number(...)`. `parseInt` is supposed to always have `radix` although i know in newer browsers it doesn't matter.

bahmutov committed:
use parseFloat

flotwig commented on packages/server/lib/browsers/index.coffee:
```diff
@@ -36,6 +37,7 @@ cleanup = ->
   instance = null
 
 getBrowserLauncherByFamily = (family) ->
+  debug("getBrowserLauncherByFamily for %s", family)
```
```suggestion
  debug("getBrowserLauncherByFamily %o", { family })
```

flotwig replied:
i find it much easier to read debug logs like this, since the actual variable name is highlighted in the log this way

flotwig commented on packages/server/lib/browsers/index.coffee:
```diff
@@ -48,9 +50,14 @@ getBrowserLauncherByFamily = (family) ->
+  findBrowsers
   .then (browsers = []) ->
+    debug("searching for browser '%s' among %s",
+      nameOrPath, pluralize("browser", browsers.length, true))
```
this seems unnecessary, debug logs should be consistent and not change wording based on what's being logged

flotwig commented on packages/server/lib/browsers/index.coffee:
```diff
@@ -48,9 +50,14 @@ getBrowserLauncherByFamily = (family) ->
+
+  findBrowsers
   .then (browsers = []) ->
+    debug("searching for browser '%s' among %s",
```
```suggestion
    debug("searching for browser %o", { nameOrPath, knownBrowsers: browsers })
```

flotwig commented on packages/server/lib/browsers/utils.coffee:
```diff
@@ -83,20 +85,24 @@ module.exports = {
     launcher.detect()
     .then (browsers = []) ->
-      version = process.versions.chrome or ""
+      debug("found %s", pluralize("browser", browsers.length, true))
```
```suggestion
      debug("found browsers %o", { browsers })
```

andrew-codes committed:
prevent jumping when mousing over values

jennifer-shehane committed:
Updated css so config doesn't jump around on hover or expand

jennifer-shehane committed:
Merge remote-tracking branch 'origin/pass-list-of-browsers-to-config-5067' into pass-list-of-browsers-to-config-5067

# Conflicts:
#	packages/desktop-gui/src/settings/settings.scss

brian-mann reviewed (changes requested):
@bahmutov
- [x] add an e2e test that mutates `config.browsers` with an object - so that the error is handled and is snapshotted
- [x] ensure that we are validating invalid `config` after its returned from the `pluginsFile`
  * added e2e tests

<img width="800" alt="Screen Shot 2019-11-15 at 3 49 54 PM" src="https://user-images.githubusercontent.com/2212006/68975042-34d96880-07c0-11ea-9c78-b7c6986c7362.png">


@andrew-codes
- [x] fix nested objects such as `browsers` jumping when you click on the arrow and expands them
- [x] add in the missing `:` colon when nested objects are expanded
- [x] show `Array(len)` when arrays are expanded in the preview area of the array
- [ ] think about maybe adding a "Copy" button that copies the derived final configuration object (non blocking, low priority)
- [x] reduce the line height to match dev tools

andrew-codes committed:
minor tweaks

- adjust spacing of items
- distinguish between arrays and objects (via expanded arrays having the text `"Array ({length})"`  after the colon

bahmutov committed:
add e2e test for catching invalid viewportWidth in cypress.json file

bahmutov committed:
validate config values returned from plugins, closes #5712

bahmutov committed:
Merge branch 'develop' into pass-list-of-browsers-to-config-5067

brian-mann committed:
Update packages/server/lib/browsers/index.coffee

Co-Authored-By: Zach Bloomquist <github@chary.us>

brian-mann committed:
Update packages/server/lib/browsers/utils.coffee

Co-Authored-By: Zach Bloomquist <github@chary.us>

brian-mann committed:
Update packages/server/lib/browsers/index.coffee

Co-Authored-By: Zach Bloomquist <github@chary.us>

bahmutov committed:
move browsers validation to regular flow

bahmutov committed:
remove using old config validation

bahmutov committed:
add plugins e2e test that returns empty browsers list

bahmutov committed:
add e2e test for invalid browser object returned from plugin

bahmutov committed:
Merge branch 'pass-list-of-browsers-to-config-5067' of github.com:cypress-io/cypress into pass-list-of-browsers-to-config-5067

bahmutov committed:
add data-driven unit tests for validating one browser object

bahmutov committed:
update unit tests

bahmutov committed:
remove moved tests

bahmutov committed:
update a few unit tests

bahmutov committed:
update integration test

bahmutov committed:
Merge branch 'develop' into pass-list-of-browsers-to-config-5067

bahmutov committed:
Merge branch 'develop' into pass-list-of-browsers-to-config-5067
